gordon/metrics/ffwd.py

```diff
@@ -0,0 +1,200 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+
+Gordon ships with a simple `ffwd <https://github.com/spotify/ffwd>`_
+metrics implementation, which can be enabled via configuration. This
+module contains the SimpleFfwdRelay, and all required classes that it
+uses to send messsages to the ffwd daemon via UDP.
+
+The `SimpleFfwdRelay` requires no configuration, but can be customized.
+The defaults that may be overridden are shown below.
+
+.. code-block:: ini
+
+    [ffwd]
+    # to identify the service creating metrics
+    key = 'gordon-service'
+
+    # the address of the ffwd daemon (see: UDPClient)
+    ip = '127.0.0.9'
+    port = 19000
+
+    # a scaling factor for timing (see: FfwdTimer)
+    time_unit = 1E9
+
+"""
+
+import asyncio
+import json
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+
+class UDPClientProtocol(asyncio.DatagramProtocol):
+    """Protocol for sending one-off messages via UDP.
+
+    Args:
+        message (bytes): Message for ffwd agent.
+    """
+    def __init__(self, message):
+        self.message = message
+        self.transport = None
+
+    def connection_made(self, transport):
+        """Create connection, use to send message and close.
+
+        Args:
+            transport (asyncio.DatagramTransport): Transport used for sending.
+        """
+        self.transport = transport
+        self.transport.sendto(self.message)
+        self.transport.close()
+
+
+class UDPClient:
+    """Client for sending UDP datagrams.
+
+    Args:
+        ip (str): (optional) Destination IP address (default: 127.0.0.1).
+        port (int): (optional) Destination port (default: 9000).
+        loop (asyncio.AbstractEventLoop impl): (optional) Event loop.
+    """
+
+    DEFAULT_IP = '127.0.0.1'
+    DEFAULT_PORT = 19000
+
+    def __init__(self, ip=None, port=None, loop=None):
+        self.ip = ip or self.DEFAULT_IP
+        self.port = port or self.DEFAULT_PORT
+        self.loop = loop or asyncio.get_event_loop()
+
+    async def send(self, metric):
+        """Transform metric to JSON bytestring and send to server.
+
+        Args:
+            metric (dict): Complete metric to send as JSON.
+        """
+        message = json.dumps(metric).encode('utf-8')
+        await self.loop.create_datagram_endpoint(
+            lambda: UDPClientProtocol(message),
+            remote_addr=(self.ip, self.port))
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class FfwdTimer:
+    """Timer which sends UDP messages to FFWD on completion.
+
+    Args:
+        metric (dict): Dict representation of the metric to send.
```

Comment:
The "dict representation of the metric" makes me think that metric could be a popo (like pojo, but Python!) just to make it more defined compared to a loose dictionary. This is just a loose thought about the design, not a request to change anything btw.

Reply:
It could be, but I think it'd be overkill for this purpose. I know there are other reasons, but generally the rule I've been using with Python is that I create an data object with attributes, it's because I know what should be in that object and want to hard fail if something tries to access it and it isn't there.  But with a [ffwd json metric](https://github.com/spotify/ffwd/tree/master/modules/json), nearly all the fields are optional, so the timer's not really counting on anything* being there, just blindly passing it through.

* you could argue we do require a 'type'='metric', but for the time being since the only thing creating timers and metrics is in the same module, I feel like it's safe enough ™.

Reply:
Thanks, good points. I've seen code that puts things like money or IPs into objects, but those cases were closer to the core logic  for those projects. I think I've been thinking more and more about stricter checks combined with type checking, so just curious to learn more what others think around this area and where the sweet spot lies.